### PR TITLE
[NanoAOD] Don't handle column types redundantly anymore

### PR DESCRIPTION
**Pull request by guitargeek**
#### PR description:

This is just a re-opening of https://github.com/cms-nanoAOD/cmssw/pull/401 from September 2019, but this time to upstream CMSSW instead of cmssw-nanoAOD. Back in the day I was interested in storing columns of so far not supported basic types in NanoAOD (like `uint16_t`), and made some developments which are I think the first steps towards this. Since https://github.com/cms-sw/cmssw/pull/30273 set the precedent that technical NanoAOD developments are acceptable outside cmssw-nanoAOD, I reopen the PR here almost a year later. Original description in the following.

----------

Hi NanoAOD devs!

I hope this is the right cmssw fork and branch for this PR.

Yesterday I wanted to introduce some new column types in my private nanoAOD productions (int16_t for example, to save a bit of space) and use them in the flat table producers. However, I realized that there are many parts of the NanoAOD code which have to be tweaked if you want to do this, as the way how column types are handled is not completely trivial.

One source of complication is that when you add a column to a flat table with addColumn(), you have to pass the type as a template argument as well as an enum value in the function parameters. After working a bit with the code, I understood that this is redundant, because the check_type function makes sure you always use the right enum value with the right template parameter. Therefore, we could just drop this enum parameter and deduce it from the template argument. In this situation, we would also not need check_type anymore.

The only tricky part are bool columns, which should actually be represented by a uint8_t vector. So far, the logic to take care of this had to be implemented in the plugins that made use of the FlatTable class, but I think I found a way to have this logic directly in the FlatTable class so one can just use addColumn<bool> to create bool columns and they will be internally stored in the uint8_t vector.

What do you think? This simplifies the type handling already quite a bit, and I think it's the good path towards a FlatTable class that will support all basic types that you can also store in TTrees.

I tested this with the local matrix tests so far, can the nano-bot tests still be done here? That would be very cool!

Thanks for considering this and cheers,
Jonas

#### PR validation:

* CMSSW compiles
* local matrix tests pass
* bot tests in cms-nanoAOD pass (see https://github.com/cms-nanoAOD/cmssw/pull/401)

#### if this PR is a backport please specify the original PR and why you need to backport that PR:

No backport  intended.

### CODE REVIEW

**Comment by slava77** [comment]
@cmsbuild please test

**Comment by silviodonato** [comment]
please test
#30440 should solve the error

**Comment by slava77** [comment]
@cmsbuild please test

**Comment by slava77** [comment]
please abort
the IB is apparently still broken

**Comment by slava77** [comment]
@cmsbuild please test

File: DataFormats/NanoAOD/interface/FlatTable.h [hunk 1]
```diff
@@ -104,40 +106,43 @@ namespace nanoaod {
...
+        uint8s_.insert(uint8s_.end(), values.begin(), values.end());
+      } else if constexpr (std::is_same<T, float>()) {
+        columns_.emplace_back(name, docString, ColumnType::Float, floats_.size());
+        floats_.insert(floats_.end(), values.begin(), values.end());
+        flatTableHelper::MaybeMantissaReduce<float>(mantissaBits).bulk(columnData<float>(columns_.size() - 1));
+      } else {
+        ColumnType type = defaultColumnType<T>();
+        auto &vec = bigVector<T>();
+        columns_.emplace_back(name, docString, type, vec.size());
+        vec.insert(vec.end(), values.begin(), values.end());
```
Comment (slava77) [inline DataFormats/NanoAOD/interface/FlatTable.h:126]: I'm missing why this if/else specialization is needed.
It looks like the code is already set up to be generic and on L126 we just need
```C++
flatTableHelper::MaybeMantissaReduce<type>(mantissaBits).bulk(columnData<type>(columns_.size() - 1));
```
(or is it "T")

**Comment by guitargeek** [comment]
Hi @slava77, thanks for your comment. I had to rebase and force-push because the code was previously based on a broken IB, but all the previous commits up to the last one remained unchanged. I implemented your comment and also found a solution to get rid of the remaining `constexpr if` in `addColumn` by moving the special treatment of the `bool` column to a more fundamental place.

Exceptions related to unsupported type usage have also been replaced with `static_assert`. The way how this is achieved might look a bit strange, but is actually the common solution cited by the cppreference:
https://en.cppreference.com/w/cpp/language/if#Constexpr_If

Local matrix tests pass.

File: DataFormats/NanoAOD/interface/FlatTable.h (hunk 1 above)
Comment (guitargeek) [reply DataFormats/NanoAOD/interface/FlatTable.h]: Indeed, you are right. I didn't look at `flatTableHelper::MaybeMantissaReduce` carefully enough.

**Comment by slava77** [comment]
@cmsbuild please test

**Comment by guitargeek** [comment]
Hi, I fixed the missing include that caused the header inconsistencies and also added some exceptions that will be thrown if someone would implement more column types but not consider them in all relevant `switch` statements. That's probably better than silently ignoring these columns.

**Comment by slava77** [comment]
@cmsbuild please test

**Comment by slava77** [comment]
@cmsbuild please test

**Comment by peruzzim** [comment]
> This is just a re-opening of [cms-nanoAOD#401](https://github.com/cms-nanoAOD/cmssw/pull/401) from September 2019, but this time to upstream CMSSW instead of cmssw-nanoAOD. Back in the day I was interested in storing columns of so far not supported basic types in NanoAOD (like `uint16_t`), and made some developments which are I think the first steps towards this. Since #30273 set the precedent that technical NanoAOD developments are acceptable outside cmssw-nanoAOD, I reopen the PR here almost a year later. 

@guitargeek I am quite uncomfortable with such a big PR being implemented outside of the cms-nanoAOD repository.
The "precedent" you mention was just touching a couple of files with changes of technical nature, yes.
This PR, instead, is modifying most of the producers with great chance of conflicts with other changes. It is indeed a good moment to do it, because of little concurrent development, but still it should be done through the cms-nanoAOD repository to ensure a full testing for all combinations of datasets and releases, and validate the result.
Also, we probably do want to backport this to avoid future backports to 10_2 and 10_6 to become a nightmare (we will need to maintain those branches for quite a long time still)

**Comment by guitargeek** [comment]
Thanks @peruzzim, I completely agree with you. We can discuss offline and once cms-nanoAOD is back in sync, we can continue there. It's nice that we can do the testing in cms-nanoAOD with the review comments by @slava77 already addressed!

**Comment by slava77** [comment]
@cmsbuild please test

File: DataFormats/NanoAOD/interface/FlatTable.h [hunk 2]
```diff
@@ -161,25 +159,30 @@ namespace nanoaod {
...
-      return bigVector<T>().begin() + col.firstIndex;
+    auto beginData(unsigned int column) {
+      return bigVector<T>().begin() + columns_[column].firstIndex;
     }
 
     template <typename T>
-    const std::vector<T> &bigVector() const {
-      throw cms::Exception("unsupported type");
+    auto const &bigVector() const {
+      CMS_SA_ALLOW return const_cast<FlatTable *>(this)->bigVector<T>();
```
Comment (slava77) [inline DataFormats/NanoAOD/interface/FlatTable.h:172]: oddly enough we are still getting 
```const_cast was used, this may result in thread-unsafe code```
https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-ce380e/7607/llvm-analysis/report-819fa2.html#EndPath

@makortel @Dr15Jones 
please advise if there is a solution to silence the SA warning

Comment (makortel) [reply DataFormats/NanoAOD/interface/FlatTable.h:172]: I suspect the silencing of `const_cast` with `CMS_SA_ALLOW` has not made from an idea to the static analyzer. We could take another look in that, but for time being the best alternative is to add comment explaining why the `const_cast` is acceptable here.

Comment (guitargeek) [reply DataFormats/NanoAOD/interface/FlatTable.h:172]: Okay! Should I still keep the `CMS_SA_ALLOW` anyway? It doesn't seem to be harmful, and if it one day also silences `const_cast` it is already there.

I can add a comment in the line before:
```
// justified use of `const_cast` to provide a `const` version of the `bigVector` accessor without code duplication
```

Comment (guitargeek) [reply DataFormats/NanoAOD/interface/FlatTable.h]: I was just checking if the core guidelines have anything to say about this situation, and actually :+1:

**Comment by slava77** [comment]
@cmsbuild please test

**Comment by slava77** [comment]
+1

for https://github.com/cms-sw/cmssw/pull/30436	f8d93aa 
- code changes are in line with the PR description and the follow up review
- jenkins tests pass and comparisons with the baseline show no differences
- from https://github.com/cms-sw/cmssw/pull/30436#issuecomment-652832213 I expect that this PR is actually not merged directly but instead goes through the cms-nanoAOD repository

**Comment by santocch** [comment]
+1

File: FWCore/Utilities/interface/Range.h [hunk 3]
```diff
@@ -16,6 +18,12 @@ namespace edm {
     T end() const { return end_; }
 
     bool empty() const { return begin_ == end_; }
+    auto size() const { return end_ - begin_; }
+
+    auto const& operator[](std::size_t idx) const { return *(begin_ + idx); }
+
+    auto const& front() const { return *begin_; }
+    auto const& back() const { return *(end_ - 1); }
```
Comment (makortel) [inline FWCore/Utilities/interface/Range.h:26]: These additions make `edm::Range` to essentially require random-access iterators. They also make `edm::Range` to look more like C++20 [`std::span`](https://en.cppreference.com/w/cpp/container/span) than the C++20 [`std::ranges::range`](https://en.cppreference.com/w/cpp/ranges/range) concept. How about leaving `edm::Range` as it is and create new `edm::Span` with these additions (that would later be replaced with `std::span`)?

Comment (guitargeek) [reply FWCore/Utilities/interface/Range.h:26]: Absolutely! I also realized that it now requires random-access iterators to work efficiently, but thought that was okay since we only use `edm::Range` for those (so far). Your idea of making it consistent with `std` is really nice, thank you. I'll do that.

**Comment by makortel** [comment]
@cmsbuild, please test

File: DataFormats/NanoAOD/interface/FlatTable.h [hunk 4]
```diff
@@ -26,7 +27,8 @@ namespace nanoaod {
       inline float one(const float &val) const {
         return (bits_ > 0 ? MiniFloatConverter::reduceMantissaToNbitsRounding(val, bits_) : val);
       }
-      inline void bulk(boost::sub_range<std::vector<float>> data) const {
+      template <typename Span>
+      inline void bulk(Span &&data) const {
```
Comment (makortel) [inline DataFormats/NanoAOD/interface/FlatTable.h:31]: Why rvalue reference instead of
```suggestion
      inline void bulk(Span const& data) const {
```
?

Comment (guitargeek) [reply DataFormats/NanoAOD/interface/FlatTable.h:31]: I once read this blog post about move semantics:
https://eli.thegreenplace.net/2014/perfect-forwarding-and-universal-references-in-c/#id6

Specifically the part "Don't let `T&&` fool you here ...". So it's not a rvalue reference, but makes the function automatically deduce if taking the argument by value or by reference is appropriate. Therefore it appeared to me that it might be a good habit to use `&&` by default for template deduced arguments, especially in this class where `bulk` is used with an rvalue.

Comment (Dr15Jones) [reply DataFormats/NanoAOD/interface/FlatTable.h:31]: `T&&` is sometimes referred to as the _universal reference_.

Comment (guitargeek) [reply DataFormats/NanoAOD/interface/FlatTable.h:31]: Maybe this isocpp blogpost is a more official reference for these universal references:
https://isocpp.org/blog/2012/11/universal-references-in-c11-scott-meyers

Comment (makortel) [reply DataFormats/NanoAOD/interface/FlatTable.h:31]: Oh right, forgot that detail, never mind then.

Comment (Dr15Jones) [reply DataFormats/NanoAOD/interface/FlatTable.h:31]: In this particular case I do not see a need for the universal reference as the type is directly interacted with instead of passed to another function via the use of `std::forward`.

Comment (guitargeek) [reply DataFormats/NanoAOD/interface/FlatTable.h:31]: That's also true

Comment (guitargeek) [reply DataFormats/NanoAOD/interface/FlatTable.h:31]: But it's such a minor change that I think I can also do it later in one go when I rebase this PR for the nanoAOD repository, or if you have more major comments, if this is okay.

Speaking about perfect forwarding, I recently noticed in the PR where I used FWCore/SOA that the SOA Table might make several unnecessary copies because even though `std::forward` is used, this is not matched by the corresponding universal reference for the argument, for example here (and also in other functions for the implementation details):
https://github.com/cms-sw/cmssw/blob/master/FWCore/SOA/interface/Table.h#L145

I didn't open an issue about this because I didn't think it was that important, but I it's just something to keep in mind for the next time anyone will work with FWCore/SOA.

Comment (makortel) [reply DataFormats/NanoAOD/interface/FlatTable.h:31]: > But it's such a minor change that I think I can also do it later in one go when I rebase this PR for the nanoAOD repository, or if you have more major comments, if this is okay.

Perfectly fine for me, it is minor indeed.

> I didn't open an issue about this because I didn't think it was that important, but I it's just something to keep in mind for the next time anyone will work with FWCore/SOA.

The best way to remind us (`core`) is to open an issue :)

Comment (guitargeek) [reply DataFormats/NanoAOD/interface/FlatTable.h:31]: Ok thanks!

**Comment by slava77** [comment]
+1

for https://github.com/cms-sw/cmssw/pull/30436	ea41702 
- changes since the last signoff are technical for reco (requested by core, to move from what was proposed to be edm::Range to become edm::Span)
- tests pass and comparisons show no differences

**Comment by makortel** [comment]
+1

**Comment by santocch** [comment]
+1

**Comment by mariadalfonso** [comment]
@guitargeek 
I'm in favour of having the technical PR merged-reviews in the cms-sw (rather than in cms-nano).

1. Can you update the,  cms-nanoAOD#401 so that I test the nano-bot
2. do you need to feature in the 10_6 as well (release that likely will be used for the Run2-nano-UL ?)

**Comment by guitargeek** [comment]
Hi @mariadalfonso, thank you very much for making a decision, it sound very good to me!

I will update the cms-nanoAOD PR and implement the latest comment from @makortel about the universal reference in the cms-sw PR so it can be merged.

About the backports: @peruzzim mentioned in his last comment (https://github.com/cms-sw/cmssw/pull/30436#issuecomment-652832213) that both the 10_2_X and 10_6_X cycles still need to be kept in sync with the recent NanoAOD developments. That would mean that this PR has to be backported to both 10_2_X and 10_6_X, no?

**Comment by guitargeek** [comment]
Once again a rebase because the IB I based my developments on expired. The only actual difference is the last added commit.

**Comment by guitargeek** [comment]
PR https://github.com/cms-nanoAOD/cmssw/pull/401 is now in sync with this PR.

**Comment by perrotta** [comment]
please test

**Comment by guitargeek** [comment]
I just tested how complicated a backport to 10_6_X would be, and it would already be a nightmare due to the clang-format PR https://github.com/cms-sw/cmssw/pull/27353. There was already some manual editing done in previous backports to circumvent the absence of clang-format changes in older releases, so at this point the clang-format commit can't be trivially backported anymore.

I think this is not a good situation, because it stalls NanoAOD developments. People are on one hand reluctant to do anything for master because the backports are so terrible, and on the other hand it's also a nightmare to "forward-port" NanoAOD developments that you did in e.g. 10_2_X to improve NanoAOD for your analysis. There are for example reported problems in the hypernews like [here]<https://hypernews.cern.ch/HyperNews/CMS/get/generators/4710.html> that everyone (also me) seems to solve for herself in 10_2_X.

In conclusion: I can't do any backports of this PR without manual changes, which would only make the situation worse. Therefore, I advise against merging my own PR before the fundamental NanoAOD backporting problem is solved. Backporting the clang-format commit might be a good start.

**Comment by slava77** [comment]
> In conclusion: I can't do any backports of this PR without manual changes, which would only make the situation worse. Therefore, I advise against merging my own PR before the fundamental NanoAOD backporting problem is solved. Backporting the clang-format commit might be a good start.

What are the benefits of backporting (even to 106X)?
Can Nano move to a more recent production release, e.g. 11_1_X?

**Comment by slava77** [comment]
+1

for https://github.com/cms-sw/cmssw/pull/30436	00b26e7 
- rebased since the last signoff for https://github.com/cms-sw/cmssw/pull/30436	ea41702 
with an addition of somewhat trivial update in 00b26e72b56c4f5661d8bb074b59edb209b3a232
- jenkins tests pass and comparisons show no differences

**Comment by makortel** [comment]
+1

**Comment by mariadalfonso** [comment]
+1

**Comment by mariadalfonso** [comment]
> > In conclusion: I can't do any backports of this PR without manual changes, which would only make the situation worse. Therefore, I advise against merging my own PR before the fundamental NanoAOD backporting problem is solved. Backporting the clang-format commit might be a good start.
> 
> What are the benefits of backporting (even to 106X)?
> Can Nano move to a more recent production release, e.g. 11_1_X?

For the analysis or dedicated production, I think the 11_1_X can be used indeed
But when we will do the UL-reminiAOD I immagine the 106X and at the same time some version of nano will be produced.

We are planning a XPOG meeting this Tuesday with a general discussion
https://indico.cern.ch/event/939863

**Comment by guitargeek** [comment]
Thank you very much!

**Comment by slava77** [comment]
> > > In conclusion: I can't do any backports of this PR without manual changes, which would only make the situation worse. Therefore, I advise against merging my own PR before the fundamental NanoAOD backporting problem is solved. Backporting the clang-format commit might be a good start.
> > 
> > 
> > What are the benefits of backporting (even to 106X)?
> > Can Nano move to a more recent production release, e.g. 11_1_X?
> 
> For the analysis or dedicated production, I think the 11_1_X can be used indeed
> But when we will do the UL-reminiAOD I immagine the 106X and at the same time some version of nano will be produced.
> 
> We are planning a XPOG meeting this Tuesday with a general discussion
> https://indico.cern.ch/event/939863

I wish we didn't have to backport so much stuff to 10_6_X for reminiAOD either, since we actually support remini in the master a more backport-friendly recent production release could be (have been) used.
At least we should minimize needs to backport to 10_2_X as well.
Let's see what comes out from the meeting discussion.

**Comment by santocch** [comment]
+1

**Comment by silviodonato** [comment]
@mariadalfonso @gouskos @peruzzim Can I merge this PR? Shall I wait for cms-nanoAOD#401 ?

**Comment by guitargeek** [comment]
Hi @silviodonato , I think Mariarosaria was pretty clear in https://github.com/cms-sw/cmssw/pull/30436#issuecomment-661719525 and especially https://github.com/cms-nanoAOD/cmssw/pull/401#issuecomment-662406510 that this PR to cms-sw is the one to merge.

**Comment by silviodonato** [comment]
+1
